### PR TITLE
chore: bump slevomat from 6.4.1 to 8.4.0

### PR DESCRIPTION
**Pull request by saas786**
CU-Task: https://app.clickup.com/t/30xcmqp

* This update brings performance to linting / formatting, at least I noticed some performance while formatting code.
* Also brings: `SlevomatCodingStandard.Complexity.Cognitive`
	ref: https://github.com/slevomat/coding-standard/commit/7a3febca6f784587aa9ea772f6de94c7cef8bbbe

---

Update:

v8.3.0
```markdown
- Added `SlevomatCodingStandard.Complexity.Cognitive` (thanks to [@bkdotcom](https://github.com/bkdotcom))
- Added `SlevomatCodingStandard.Files.FileLength` (thanks to [@bkdotcom](https://github.com/bkdotcom))
- Added `SlevomatCodingStandard.Classes.ClassLength` (thanks to [@bkdotcom](https://github.com/bkdotcom))
```

v8.2.0
```markdown
- Added `SlevomatCodingStandard.Classes.BackedEnumTypeSpacing`
```

v8.0.0

```markdown
## ⚠️BC breaks

- `SlevomatCodingStandard.TypeHints.PropertyTypeHintSpacing` renamed to `SlevomatCodingStandard.Classes.PropertyDeclaration`
- `SlevomatCodingStandard.Classes.ClassStructure`: Removed option `enableFinalMethods`
- Removed error `SlevomatCodingStandard.Namespaces.UnusedUses.MismatchingCaseSensitivity`
```

v7.2.0
```markdown
-   Added `SlevomatCodingStandard.Functions.DisallowTrailingCommaInClosureUse`
-   Added `SlevomatCodingStandard.Functions.RequireTrailingCommaInClosureUse`
```

v7.1.0
```markdown
- Added `SlevomatCodingStandard.Classes.RequireAbstractOrFinal` (thanks to [@roslov](https://github.com/roslov))
- Added `SlevomatCodingStandard.Exceptions.DisallowNonCapturingCatch` (thanks to [@olsavmic](https://github.com/olsavmic))
```

v7.0.0

```markdown
-   `RequireNonCapturingCatchSniff`: Requires non-capturing `catch` when the variable with exception is not used
-   `RequireNullSafeObjectOperatorSniff`: Requires using `?->` operator
-   `DisallowNullSafeObjectOperatorSniff`: Disallows using `?->` operator
-   `RequireTrailingCommaInDeclarationSniff`: Enforces trailing commas in multi-line declarations
-   `DisallowTrailingCommaInDeclarationSniff`: Disallows trailing commas in multi-line declarations
-   `RequireConstructorPropertyPromotionSniff`: Requires use of constructor property promotion
-   `DisallowConstructorPropertyPromotionSniff`: Disallows usage of constructor property promotion
-   `UnionTypeHintFormatSniff`: Checks format of union type hints
-   `DisallowNamedArgumentsSniff`: Disallows usage of named arguments
-   `DisallowTrailingCommaInCallSniff`: Disallows trailing commas in multi-line calls
-   `ForbiddenPublicPropertySniff`: Disallows using public properties (thanks to [@50bhan](https://github.com/50bhan))
-   `FunctionLengthSniff`: Disallows long function (thanks to [@50bhan](https://github.com/50bhan))

## ⚠️BC breaks

-   `PropertyTypeHintSpacingSniff`: Removed `CODE_NO_SPACE_BEFORE_TYPE_HINT` because it's not possible on PHP 8
-   `DeclareStrictTypesSniff`: Options renamed and modifed, see [README](https://github.com/slevomat/coding-standard#slevomatcodingstandardtypehintsdeclarestricttypes-)
-   `TrailingCommaInCallSniff` renamed to `RequireTrailingCommaInCallSniff`
-   Removed deprecated `UnusedPrivateElementsSniff`
-   Removed `FullyQualifiedClassNameAfterKeywordSniff`
-   `ReferenceUsedNamesOnlySniff`: Removed option `fullyQualifiedKeywords`
-   `BlockControlStructureSpacingSniff`: Options renamed and modified, see [README](https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresblockcontrolstructurespacing-)
-   `JumpStatementsSpacingSniff`: Options renamed and modified, see [README](https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresjumpstatementsspacing-)
-   `ParentCallSpacingSniff`: Shorter options names, see [README](https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesparentcallspacing-)
-   All helpers are marked as internal
-   `UselessFunctionDocCommentSniff`: Removed suppress support
```

### CODE REVIEW

**Review by pawelkmpt — COMMENTED** [review]
@saas786 since you pushed this PR Slevomat got even newer version 8.2.0. Maybe we should bump it directly to that? Can you check?

Also, when rule is changed it'd be good to have information why and what we'll get from it.

**Comment by saas786** [comment]
> @saas786 since you pushed this PR Slevomat got even newer version 8.2.0. Maybe we should bump it directly to that? Can you check?
> 
> Also, when rule is changed it'd be good to have information why and what we'll get from it.

Sure, can give v8 a go in this PR.

Rule is renamed in this case. No longer available in v7, new named rule is used instead.

**Comment by lkraav** [comment]
I postponed bumping Slevomat, because we couldn't make our codebases compatible even with v6, so felt a little pointless adding even more work on top that's likely really difficult to finish.

Maybe if everyone did a single rule / week, it would be done in a few months?

**Comment by saas786** [comment]
> I postponed bumping Slevomat, because we couldn't make our codebases compatible even with v6, so felt a little pointless adding even more work on top that's likely really difficult to finish.
> 
> Maybe if everyone did a single rule / week, it would be done in a few months?

We are already applying formatting slowly to current codebase, after upgrade it should be not different.

I will update it to v8 and see which new rules are added, initially will exclude those and merge this PR.
And than we can re-include rules as per our need iteratively / gradually.

Make sense?

**Comment by pawelkmpt** [comment]
Approving this depends on https://github.com/conversionxl/institute-plugin/pull/199 discussions

**Comment by saas786** [comment]
> Approving this depends on [conversionxl/institute-plugin#199](https://github.com/conversionxl/institute-plugin/pull/199) discussions

`conversionxl/institute-plugin#199` should be discussed independently.

This should be merged, so we can later tweak rules as per our needs.
Whenever we do `phpcbf` on some code and you feel it has applied some rule, which we should not apply to our code, hence we can specifically mark that rule.

**Comment by pawelkmpt** [comment]
> This should be merged, so we can later tweak rules as per our needs.
> Whenever we do `phpcbf` on some code and you feel it has applied some rule, which we should not apply to our code, hence we can specifically mark that rule.

If I merge it, I need to:
1. update package in `institute-plugin`
2. run `phpcbf` in `institute-plugin`
3. find out code is modified the way I don't like in 60 places
4. adjust ruleset

and repeat these steps until desired result in formatting of `institute-plugin` code is achieved.
So there's a risk of having multiple ruleset update commits in plugin code spamming git log.

I believe it's better to adjust this ruleset locally, run `phpcbf` and once we get to the point when we're satisfied with the code, merge this and bump package in `institute-plugin`.

All of that with an assumption that this new version of Slevomat caused changes seen in the https://github.com/conversionxl/institute-plugin/pull/199

**Comment by saas786** [comment]
> If I merge it, I need to:
> 
> 1. update package in `institute-plugin`
> 2. run `phpcbf` in `institute-plugin`
> 3. find out code is modified the way I don't like in 60 places
> 4. adjust ruleset

I am suggesting:

1) We merge this PR.
2) We run `phpcbf` on any new PR (changed files only), like we already do: [`a8cd8ef` (#196)](https://github.com/conversionxl/institute-plugin/pull/196/commits/a8cd8ef1f57a87ee8aeccc27c8d82c277d0fc682)
3) If it applies some formatting which we don't like, we inspect relevant rule and tweak / remove it.

Ignore `conversionxl/institute-plugin#199`, that is not relevant to this PR, that I only filled in case we are interested in single PR to perform all formatting changes, and also to review if `slevomat/coding-standard:^8.3` brought any significant changes.

> All of that with an assumption that this new version of Slevomat caused changes seen in the [conversionxl/institute-plugin#199](https://github.com/conversionxl/institute-plugin/pull/199)

These changes are not caused by `slevomat/coding-standard:^8.3`, if you don't use this PR, and instead use master branch, you should see almost all the changes reflected in `conversionxl/institute-plugin#199`. So its not specific to this PR.

File: CXL/ruleset.xml [hunk 1]
```diff
@@ -124,29 +134,50 @@
...
         <exclude name="SlevomatCodingStandard.ControlStructures.DisallowShortTernaryOperator.DisallowedShortTernaryOperator"/>
         <exclude name="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
         <exclude name="SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing"/>
+        <!-- There are two rules which conflict. NewWithoutParentheses and UselessParentheses. One must be disabled -->
+        <!-- We allow new Class(); rather than new Class;-->
         <exclude name="SlevomatCodingStandard.ControlStructures.NewWithoutParentheses"/>
         <exclude name="SlevomatCodingStandard.ControlStructures.NewWithParentheses.MissingParentheses"/>
+        <!-- Disabled by default, but you may be interested in reading up on Yoda conditions to see if it's -->
+        <!-- something that you would like to start using: https://en.wikipedia.org/wiki/Yoda_conditions -->
         <exclude name="SlevomatCodingStandard.ControlStructures.RequireYodaComparison"/>
```
Comment (pawelkmpt) [inline CXL/ruleset.xml:161]: We want yoda conditions, but I think they're covered in WP coding standards.

File: CXL/ruleset.xml [hunk 2]
```diff
@@ -124,29 +134,50 @@
...
         <exclude name="SlevomatCodingStandard.Files.TypeNameMatchesFileName"/>
         <exclude name="SlevomatCodingStandard.Functions.DisallowArrowFunction.DisallowedArrowFunction"/>
+        <!-- Length does not determine complexity. We do not care if methods or files are long -->
+        <exclude name="SlevomatCodingStandard.Functions.FunctionLength.FunctionLength"/>
+        <!-- Also this is deprecated -->
+        <exclude name="SlevomatCodingStandard.Files.FunctionLength.FunctionLength"/>
         <exclude name="SlevomatCodingStandard.Functions.RequireMultiLineCall"/>
-        <exclude name="SlevomatCodingStandard.Functions.TrailingCommaInCall"/>
+        <!-- We do not require trailing commas on multiline methods. Both rules disabled, as we'll allow folks to -->
+        <!-- add them if they want them, but we don't enforce either way -->
```
Comment (pawelkmpt) [inline CXL/ruleset.xml]:
```suggestion
        <!-- We do not require trailing commas on multiline methods. -->
```

File: CXL/ruleset.xml [hunk 3]
```diff
@@ -124,29 +134,50 @@
...
+        <exclude name="SlevomatCodingStandard.Functions.RequireTrailingCommaInCall"/>
+        <!-- Global constants and exceptions do not need to be fully qualified -->
         <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedGlobalConstants.NonFullyQualified"/>
         <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedGlobalFunctions.NonFullyQualified"/>
+        <!-- We generally allow the use of any namespace -->
         <exclude name="SlevomatCodingStandard.Namespaces.UseOnlyWhitelistedNamespaces"/>
+        <!-- Inline doc comments are fine. We use them all the time to describe things like $dataList->first() -->
```
Comment (pawelkmpt) [inline CXL/ruleset.xml]:
```suggestion
        <!-- Inline doc comments are fine. We use them quite often -->
```

**Review by pawelkmpt — CHANGES_REQUESTED** [review]
@saas786 when you get inspirations from 3rd party do not blindly copy everything, adjust it for us.

In general exclusions look good. I'm gonna test it on our codebase now.

**Comment by pawelkmpt** [comment]
I cloned this repo and switched to relevant branch. Then did `composer i`. I get this error:

```
./vendor/bin/phpcs -s src/Plugin/Integrations/*
ERROR: Referenced sniff "SlevomatCodingStandard.Classes.RequireAbstractOrFinal" does not exist
```

EDIT

Same for:

* `SlevomatCodingStandard.Functions.RequireTrailingCommaInCall`
* `SlevomatCodingStandard.Complexity.Cognitive`


EDIT 2

I see these though in results:
* `SlevomatCodingStandard.Functions.TrailingCommaInCall.MissingTrailingComma`

**Comment by pawelkmpt** [comment]
I looked at [one of the resources](https://github.com/andrewandante/silverstripe-async-publisher/blob/v1.1.0/phpcs.xml.dist#L8-L11) @saas786 linked in commit message and I like it:

```xml
    <!-- Show progress and output sniff names on violation, and add colours -->
    <arg value="p" />
    <arg name="colors" />
    <arg value="s" />
```

@lkraav WDYT?

Screenshot 2022-08-23 at 17 23 43

Screenshot 2022-08-23 at 17 23 27

**Comment by pawelkmpt** [comment]
I'm wondering if these would be better as a warning 🤔 

```
 698 | ERROR   | [x] Method \ConversionXL\Institute\Plugin\Integrations\WooCommerce\Memberships::get_user_active_team_ids() has useless @param annotation for parameter
     |         |     $user_id. (SlevomatCodingStandard.TypeHints.ParameterTypeHint.UselessAnnotation)
```

```
 138 | ERROR | [x] Use arrow function. (SlevomatCodingStandard.Functions.RequireArrowFunction.RequiredArrowFunction)
```

```
 555 | ERROR | [ ] Use of super global variable is disallowed. (SlevomatCodingStandard.Variables.DisallowSuperGlobalVariable.DisallowedSuperGlobalVariable)
```

File: CXL/ruleset.xml (hunk 1 above)
Comment (saas786) [reply CXL/ruleset.xml:161]: > We want yoda conditions, but I think they're covered in WP coding standards.

I didn't exclude it, it was excluded already. I just added comment.

See possible reasoning: https://en.wikipedia.org/wiki/Yoda_conditions#Criticism
Also: https://github.com/slevomat/coding-standard#slevomatcodingstandardcontrolstructuresdisallowyodacomparisonrequireyodacomparison-

**Comment by saas786** [comment]
> I cloned this repo and switched to relevant branch. Then did `composer i`. I get this error:

Updated it inside `institute plugin` as well?

You need to update it inside `institute plugin`.

Ref: https://github.com/conversionxl/institute-plugin/blob/master/composer.json#L37

I use https://github.com/franzliedke/studio, makes it easier to use local copy of some repo.

`studio load '../phpcs-ruleset'`
`composer update cxl/phpcs-ruleset -W`

**Comment by saas786** [comment]
> I'm wondering if these would be better as a warning 🤔
> 
> ```
>  698 | ERROR   | [x] Method \ConversionXL\Institute\Plugin\Integrations\WooCommerce\Memberships::get_user_active_team_ids() has useless @param annotation for parameter
>      |         |     $user_id. (SlevomatCodingStandard.TypeHints.ParameterTypeHint.UselessAnnotation)
> ```
> 
> ```
>  138 | ERROR | [x] Use arrow function. (SlevomatCodingStandard.Functions.RequireArrowFunction.RequiredArrowFunction)
> ```
> 
> ```
>  555 | ERROR | [ ] Use of super global variable is disallowed. (SlevomatCodingStandard.Variables.DisallowSuperGlobalVariable.DisallowedSuperGlobalVariable)
> ```

I think they are fine as is.

1) `SlevomatCodingStandard.TypeHints.ParameterTypeHint.UselessAnnotation` will be fixed auto.
2) `SlevomatCodingStandard.Functions.RequireArrowFunction.RequiredArrowFunction`, will be fixed auto, if we don't want to use it for some code, can do `// phpcs:ignore SlevomatCodingStandard.Functions.RequireArrowFunction.RequiredArrowFunction`
3) It should be error, as it not encouraged and can result into issue if used carelessly, so when it does throw error, we can exclude it if needed, or rewrite code to avoid it.

**Comment by pawelkmpt** [comment]
> > I cloned this repo and switched to relevant branch. Then did `composer i`. I get this error:
> 
> Updated it inside `institute plugin` as well?
> 
> You need to update it inside `institute plugin`.
> 
> Ref: https://github.com/conversionxl/institute-plugin/blob/master/composer.json#L37

I cloned directly into `/vendor/cxl/phpcs-ruleset`, did `composer i` inside of this directory.
Later I did `composer i` in main plugin dir as well, tried also without it but didn't help. It still fails.


EDIT

@saas786 I think you need to commit `composer.lock` too. I have Slevomat 8.4.0 but `composer.json` says `"slevomat/coding-standard": "^8.3"`


EDIT 2

I performed `composer u cxl/phpcs-ruleset -W` in plugin dir and now it works well 👍

**Comment by saas786** [comment]
> @saas786 I think you need to commit `composer.lock` too. I have Slevomat 8.4.0 but `composer.json` says `"slevomat/coding-standard": "^8.3"`

Lets update it too `Slevomat 8.4.0` while we are on it? Shouldn't be a big update.

**Comment by pawelkmpt** [comment]
> Lets update it too `Slevomat 8.4.0` while we are on it? Shouldn't be a big update.

Sure. Please update, commit `composer.lock` and also add these flags I mentioned above.

**Comment by saas786** [comment]
> > Lets update it too `Slevomat 8.4.0` while we are on it? Shouldn't be a big update.
> 
> Sure. Please update, commit `composer.lock` and also add these flags I mentioned above.

Done, updated to `v8.4.0` and added  `composer.lock` as well.

**Comment by pawelkmpt** [comment]
> Done, updated to `v8.4.0` and added `composer.lock` as well.

Works well. What about adding

```xml
    <!-- Show progress and output sniff names on violation, and add colours -->
    <arg value="p" />
    <arg name="colors" />
    <arg value="s" />
```

**Review by pawelkmpt — APPROVED** [review]

**Pull request merged**